File: isucon14/webapp/go/payment_gateway.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"time"
)

var erroredUpstream = errors.New("errored upstream")

type paymentGatewayPostPaymentRequest struct {
	Amount int `json:"amount"`
}

type paymentGatewayGetPaymentsResponseOne struct {
	Amount int    `json:"amount"`
	Status string `json:"status"`
}

func requestPaymentGatewayPostPayment(ctx context.Context, paymentGatewayURL string, token string, param *paymentGatewayPostPaymentRequest, retrieveRidesOrderByCreatedAtAsc func() ([]Ride, error)) error {
	b, err := json.Marshal(param)
	if err != nil {
		return err
	}

	// 失敗したらとりあえずリトライ
	// FIXME: 社内決済マイクロサービスのインフラに異常が発生していて、同時にたくさんリクエストすると変なことになる可能性あり
	retry := 0
	for {
		err := func() error {
			req, err := http.NewRequestWithContext(ctx, http.MethodPost, paymentGatewayURL+"/payments", bytes.NewBuffer(b))
			if err != nil {
				return err
			}
			req.Header.Set("Content-Type", "application/json")
			req.Header.Set("Authorization", "Bearer "+token)

			res, err := http.DefaultClient.Do(req)
			if err != nil {
				return err
			}
			defer res.Body.Close()

			if res.StatusCode != http.StatusNoContent {
				// エラーが返ってきても成功している場合があるので、社内決済マイクロサービスに問い合わせ
				getReq, err := http.NewRequestWithContext(ctx, http.MethodGet, paymentGatewayURL+"/payments", bytes.NewBuffer([]byte{}))
				if err != nil {
					return err
				}
				getReq.Header.Set("Authorization", "Bearer "+token)

				getRes, err := http.DefaultClient.Do(getReq)
				if err != nil {
					return err
				}
				defer res.Body.Close()

				// GET /payments は障害と関係なく200が返るので、200以外は回復不能なエラーとする
				if getRes.StatusCode != http.StatusOK {
					return fmt.Errorf("[GET /payments] unexpected status code (%d)", getRes.StatusCode)
				}
				var payments []paymentGatewayGetPaymentsResponseOne
				if err := json.NewDecoder(getRes.Body).Decode(&payments); err != nil {
					return err
				}

				rides, err := retrieveRidesOrderByCreatedAtAsc()
				if err != nil {
					return err
				}

				if len(rides) != len(payments) {
					return fmt.Errorf("unexpected number of payments: %d != %d. %w", len(rides), len(payments), erroredUpstream)
				}

				return nil
			}
			return nil
		}()
		if err != nil {
			if retry < 5 {
				retry++
				time.Sleep(100 * time.Millisecond)
				continue
			} else {
				return err
			}
		}
		break
	}

	return nil
}

File: isucon14/webapp/nodejs/src/payment_gateway.ts
import { ErroredUpstream } from "./common.js";
import type { Ride } from "./types/models.js";
import { setTimeout } from "node:timers/promises";

type PaymentGatewayPostPaymentRequest = {
  amount: number;
};

export const requestPaymentGatewayPostPayment = async (
  paymentGatewayURL: string,
  token: string,
  param: PaymentGatewayPostPaymentRequest,
  retrieveRidesOrderByCreatedAtAsc: () => Promise<Ride[]>,
): Promise<ErroredUpstream | Error | undefined> => {
  // 失敗したらとりあえずリトライ
  // FIXME: 社内決済マイクロサービスのインフラに異常が発生していて、同時にたくさんリクエストすると変なことになる可能性あり
  let retry = 0;
  while (true) {
    try {
      const res = await fetch(`${paymentGatewayURL}/payments`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify(param),
      });

      if (res.status !== 204) {
        // エラーが返ってきても成功している場合があるので、社内決済マイクロサービスに問い合わせ
        const getRes = await fetch(`${paymentGatewayURL}/payments`, {
          method: "GET",
          headers: {
            Authorization: `Bearer ${token}`,
          },
        });

        // GET /payments は障害と関係なく200が返るので、200以外は回復不能なエラーとする
        if (getRes.status !== 200) {
          return new Error(
            `[GET /payments] unexpected status code (${getRes.status})`,
          );
        }
        const payments = await getRes.json();

        const rides = await retrieveRidesOrderByCreatedAtAsc();
        if (rides.length !== payments.length) {
          return new ErroredUpstream(
            `unexpected number of payments: ${rides.length} != ${payments.length}`,
          );
        }
      }
      break;
    } catch (err) {
      if (retry < 5) {
        retry++;
        await setTimeout(100);
      } else {
        throw err;
      }
    }
  }
};
